jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -89,6 +90,39 @@ public class InternalDowncallHandler {
 	private MemoryLayout realReturnLayout;
 	private MethodHandle boundMH;
 
+	/*[IF JAVA_SPEC_VERSION >= 22]*/
+	/* The identifier denotes the passed-in argument is a heap segment which helps
+	 * extract the address from the heap argument's base object in native.
+	 */
+	private static final long DOWNCALL_HEAP_ARGUMENT_ID = 0x1;
+
+	/* The ThreadLocal holds the information of the heap argument which includes
+	 * the base object array, the offset array plus the current heap argument index
+	 * for the current thread in multithreading. The heap base array stores the base
+	 * object (which contains the heap address as defined in OpenJDK) of the heap
+	 * arguments to extract the heap address from the base object in native.
+	 */
+	private static final class HeapArgInfo {
+		final Object[] bases;
+		final long[] offsets;
+		int index;
+
+		HeapArgInfo(int size) {
+			bases = new Object[size];
+			offsets = new long[size];
+			index = 0;
+		}
+
+		void append(Object base, long offset) {
+			bases[index] = base;
+			offsets[index] = offset;
+			index += 1;
+		}
+	}
+
+	private final ThreadLocal<HeapArgInfo> heapArgInfo;
```

Comment:
I hope one day we can arrange to avoid this (relatively expensive) mechanism by computing native addresses earlier.

Reply:
Actually it depends on how OpenJDK compute the native address which is indirectly based on the GC related code in the scenarios with different GC policies enabled. Ideally we just call a java method to get the native address ready for use instead of adding a bunch of extra code to deal with that.

FYI: @dmitripivkine